test/dynamo/test_dynamo.py

```diff
@@ -218,72 +249,92 @@ def forward(self, index, copy_tensor, input_tensor, op_name):
-    compiled_model = torch.compile(xla_model, backend='openxla')
+    device_model = TestModel(device).to(device)
+    compiled_model = torch.compile(device_model, backend='openxla')
 
     input_tensor = torch.ones(3)
     copy_tensor = torch.rand(5, 3)
     index = torch.tensor([0, 4, 2, 1, 3])
-    xla_input_tensor = input_tensor.to(device)
-    xla_copy_tensor = copy_tensor.to(device)
-    xla_index = index.to(device)
+    device_input_tensor = input_tensor.to(device)
+    device_copy_tensor = copy_tensor.to(device)
+    device_index = index.to(device)
 
     in_place_ops = ['copy_', 'add_', 'abs_']
     for in_place_op in in_place_ops:
       res_cpu = cpu_model.forward(
           index, copy_tensor, input_tensor, op_name=in_place_op)
-      res_xla_dynamo = compiled_model.forward(
-          xla_index, xla_copy_tensor, xla_input_tensor, op_name=in_place_op)
-      self.assertTrue(torch.allclose(res_cpu, res_xla_dynamo.cpu()))
-
-  def test_einsum(self):
+      res_device_dynamo = compiled_model.forward(
+          device_index,
+          device_copy_tensor,
+          device_input_tensor,
+          op_name=in_place_op)
+      self.assertTrue(torch.allclose(res_cpu, res_device_dynamo.cpu()))
+
+  @parameterized.parameters(
+      True,
+      False,
+  )
+  def test_einsum(self, initialize_on_cuda):
     # einsum currently does not have meta function to compute the shape hence
     # will fallback to XLA with FakeTensor as input to infer the output shape.
     def einsum_mm(a, b):
       return torch.einsum('ijkl,ijlm->ijkm', a, b)
 
-    device = xm.xla_device()
-    a = torch.randn(4, 4, 4, 4).to(xm.xla_device())
-    b = torch.randn(4, 4, 4, 4).to(xm.xla_device())
+    device = self._choose_proper_device(initialize_on_cuda)
+    a = torch.randn(4, 4, 4, 4).to(device)
+    b = torch.randn(4, 4, 4, 4).to(device)
     xm.mark_step()
 
     dynamo_einsum_mm = torch.compile(einsum_mm, backend="openxla")
-    res_xla_dynamo = dynamo_einsum_mm(a, b)
-    res_xla_non_dynamo = einsum_mm(a, b)
+    res_device_dynamo = dynamo_einsum_mm(a, b)
+    res_device_non_dynamo = einsum_mm(a, b)
     self.assertTrue(
-        torch.allclose(res_xla_non_dynamo.cpu(), res_xla_dynamo.cpu()))
+        torch.allclose(res_device_non_dynamo.cpu(), res_device_dynamo.cpu()))
 
-  def test_simple_model_with_different_input_shape(self):
-    met.clear_counters()
-    device = xm.xla_device()
-    xla_x = torch.randn(5, 5).to(device)
-    xla_y = torch.randn(5, 5).to(device)
-    xla_z = torch.randn(10, 10).to(device)
+  @parameterized.parameters(
+      True,
+      False,
+  )
+  def test_simple_model_with_different_input_shape(self, initialize_on_cuda):
+    met.clear_all()
+    device = self._choose_proper_device(initialize_on_cuda)
+    # We need to make `dim` depend on `initialize_on_cuda` because the compilation cache
+    # does not clean itself between the parameterized tests.
+    dim = 5 + int(initialize_on_cuda)
```

Comment:
Do you mean (i) dynamo's cache or (ii) XLA cache? 

- If it's (i), we can just reset it
- If it's (ii), can't we just  reset it, too? If not, I would argue that we don't need to worry about it, since that's not what's being tested, here

What do you think?

Reply:
It's (ii) and I don't think there exists a way to reset the XLA compilation cache afaik. 

> I would argue that we don't need to worry about it, since that's not what's being tested, here

Actually, the existing test https://github.com/pytorch/xla/blob/ea2a6f774299abea15450670049429655829784c/test/dynamo/test_dynamo.py#L276 tests the compilation cache. Without the change, the test will fail because of the reason in the comment. That's why I have to change it..

Reply:
Ah, I see. I guess that works. You could also only run this check in one of the runs (e.g. if `initialize_on_cuda == True`). But, I think this is also ok.